File: airbyte-persistence/job-persistence/src/main/java/io/airbyte/persistence/job/tracker/TrackingMetadata.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.persistence.job.tracker;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableMap.Builder;
import io.airbyte.commons.json.Jsons;
import io.airbyte.config.ActorDefinitionVersion;
import io.airbyte.config.Attempt;
import io.airbyte.config.AttemptFailureSummary;
import io.airbyte.config.FailureReason;
import io.airbyte.config.Job;
import io.airbyte.config.JobOutput;
import io.airbyte.config.ResourceRequirements;
import io.airbyte.config.ScheduleData;
import io.airbyte.config.StandardDestinationDefinition;
import io.airbyte.config.StandardSourceDefinition;
import io.airbyte.config.StandardSync;
import io.airbyte.config.StandardSync.ScheduleType;
import io.airbyte.config.StandardSyncSummary;
import io.airbyte.config.SyncStats;
import io.airbyte.config.helpers.ScheduleHelpers;
import io.micronaut.core.util.StringUtils;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

/**
 * Helpers to fetch stats / metadata about Airbyte domain models and turn them into flat maps that
 * can be appended to telemetry calls.
 */
public class TrackingMetadata {

  /**
   * Map sync metadata / stats to a string-to-object map so it can be attached to telemetry calls.
   *
   * @param standardSync connection
   * @return metadata / stats as a string-to-object map.
   */
  public static Map<String, Object> generateSyncMetadata(final StandardSync standardSync) {
    final Builder<String, Object> metadata = ImmutableMap.builder();
    metadata.put("connection_id", standardSync.getConnectionId());
    metadata.put("source_id", standardSync.getSourceId());
    metadata.put("destination_id", standardSync.getDestinationId());

    final String frequencyString;
    if (standardSync.getScheduleType() != null) {
      frequencyString = getFrequencyStringFromScheduleType(standardSync.getScheduleType(), standardSync.getScheduleData());
    } else if (standardSync.getManual()) {
      frequencyString = "manual";
    } else {
      final long intervalInMinutes = TimeUnit.SECONDS.toMinutes(ScheduleHelpers.getIntervalInSecond(standardSync.getSchedule()));
      frequencyString = intervalInMinutes + " min";
    }
    metadata.put("frequency", frequencyString);

    final int operationCount = standardSync.getOperationIds() != null ? standardSync.getOperationIds().size() : 0;
    metadata.put("operation_count", operationCount);
    if (standardSync.getNamespaceDefinition() != null) {
      metadata.put("namespace_definition", standardSync.getNamespaceDefinition());
    }

    final boolean isUsingPrefix = standardSync.getPrefix() != null && !standardSync.getPrefix().isBlank();
    metadata.put("table_prefix", isUsingPrefix);

    final ResourceRequirements resourceRequirements = standardSync.getResourceRequirements();

    if (resourceRequirements != null) {
      if (!com.google.common.base.Strings.isNullOrEmpty(resourceRequirements.getCpuRequest())) {
        metadata.put("sync_cpu_request", resourceRequirements.getCpuRequest());
      }
      if (!com.google.common.base.Strings.isNullOrEmpty(resourceRequirements.getCpuLimit())) {
        metadata.put("sync_cpu_limit", resourceRequirements.getCpuLimit());
      }
      if (!com.google.common.base.Strings.isNullOrEmpty(resourceRequirements.getMemoryRequest())) {
        metadata.put("sync_memory_request", resourceRequirements.getMemoryRequest());
      }
      if (!com.google.common.base.Strings.isNullOrEmpty(resourceRequirements.getMemoryLimit())) {
        metadata.put("sync_memory_limit", resourceRequirements.getMemoryLimit());
      }
    }
    return metadata.build();
  }

  /**
   * Map destination definition metadata / stats to a string-to-object map so it can be attached to
   * telemetry calls.
   *
   * @param destinationDefinition destination definition
   * @return metadata / stats as a string-to-object map.
   */
  public static Map<String, Object> generateDestinationDefinitionMetadata(final StandardDestinationDefinition destinationDefinition,
                                                                          final ActorDefinitionVersion destinationVersion) {
    final Builder<String, Object> metadata = ImmutableMap.builder();
    metadata.put("connector_destination", destinationDefinition.getName());
    metadata.put("connector_destination_definition_id", destinationDefinition.getDestinationDefinitionId());
    metadata.putAll(generateActorDefinitionVersionMetadata("connector_destination_", destinationVersion));
    return metadata.build();
  }

  /**
   * Map source definition metadata / stats to a string-to-object map so it can be attached to
   * telemetry calls.
   *
   * @param sourceDefinition source definition
   * @return metadata / stats as a string-to-object map.
   */
  public static Map<String, Object> generateSourceDefinitionMetadata(final StandardSourceDefinition sourceDefinition,
                                                                     final ActorDefinitionVersion sourceVersion) {
    final Builder<String, Object> metadata = ImmutableMap.builder();
    metadata.put("connector_source", sourceDefinition.getName());
    metadata.put("connector_source_definition_id", sourceDefinition.getSourceDefinitionId());
    metadata.putAll(generateActorDefinitionVersionMetadata("connector_source_", sourceVersion));
    return metadata.build();
  }

  private static Map<String, Object> generateActorDefinitionVersionMetadata(final String metaPrefix, final ActorDefinitionVersion sourceVersion) {
    final Builder<String, Object> metadata = ImmutableMap.builder();
    metadata.put(metaPrefix + "docker_repository", sourceVersion.getDockerRepository());
    final String imageTag = sourceVersion.getDockerImageTag();
    if (!StringUtils.isEmpty(imageTag)) {
      metadata.put(metaPrefix + "version", imageTag);
    }
    return metadata.build();

  }

  /**
   * Map job metadata / stats to a string-to-object map so it can be attached to telemetry calls.
   *
   * @param job job
   * @return metadata / stats as a string-to-object map.
   */
  public static Map<String, Object> generateJobAttemptMetadata(final Job job) {
    final Builder<String, Object> metadata = ImmutableMap.builder();
    // Early returns in case we're missing the relevant stats.
    if (job == null) {
      return metadata.build();
    }
    final List<Attempt> attempts = job.getAttempts();
    if (attempts == null || attempts.isEmpty()) {
      return metadata.build();
    }
    final Attempt lastAttempt = attempts.getLast();
    if (lastAttempt.getOutput() == null || lastAttempt.getOutput().isEmpty()) {
      return metadata.build();
    }
    final JobOutput jobOutput = lastAttempt.getOutput().get();
    if (jobOutput.getSync() == null) {
      return metadata.build();
    }
    final StandardSyncSummary syncSummary = jobOutput.getSync().getStandardSyncSummary();
    final SyncStats totalStats = syncSummary.getTotalStats();
    if (syncSummary.getStartTime() != null) {
      metadata.put("sync_start_time", syncSummary.getStartTime());
    }
    if (syncSummary.getEndTime() != null && syncSummary.getStartTime() != null) {
      metadata.put("duration", Math.round((syncSummary.getEndTime() - syncSummary.getStartTime()) / 1000.0));
    }
    if (syncSummary.getBytesSynced() != null) {
      metadata.put("volume_mb", syncSummary.getBytesSynced());
    }
    if (syncSummary.getRecordsSynced() != null) {
      metadata.put("volume_rows", syncSummary.getRecordsSynced());
    }
    if (totalStats.getSourceStateMessagesEmitted() != null) {
      metadata.put("count_state_messages_from_source", syncSummary.getTotalStats().getSourceStateMessagesEmitted());
    }
    if (totalStats.getDestinationStateMessagesEmitted() != null) {
      metadata.put("count_state_messages_from_destination", syncSummary.getTotalStats().getDestinationStateMessagesEmitted());
    }
    if (totalStats.getMaxSecondsBeforeSourceStateMessageEmitted() != null) {
      metadata.put("max_seconds_before_source_state_message_emitted",
          totalStats.getMaxSecondsBeforeSourceStateMessageEmitted());
    }
    if (totalStats.getMeanSecondsBeforeSourceStateMessageEmitted() != null) {
      metadata.put("mean_seconds_before_source_state_message_emitted",
          totalStats.getMeanSecondsBeforeSourceStateMessageEmitted());
    }
    if (totalStats.getMaxSecondsBetweenStateMessageEmittedandCommitted() != null) {
      metadata.put("max_seconds_between_state_message_emit_and_commit",
          totalStats.getMaxSecondsBetweenStateMessageEmittedandCommitted());
    }
    if (totalStats.getMeanSecondsBetweenStateMessageEmittedandCommitted() != null) {
      metadata.put("mean_seconds_between_state_message_emit_and_commit",
          totalStats.getMeanSecondsBetweenStateMessageEmittedandCommitted());
    }

    if (totalStats.getReplicationStartTime() != null) {
      metadata.put("replication_start_time", totalStats.getReplicationStartTime());
    }
    if (totalStats.getReplicationEndTime() != null) {
      metadata.put("replication_end_time", totalStats.getReplicationEndTime());
    }
    if (totalStats.getSourceReadStartTime() != null) {
      metadata.put("source_read_start_time", totalStats.getSourceReadStartTime());
    }
    if (totalStats.getSourceReadEndTime() != null) {
      metadata.put("source_read_end_time", totalStats.getSourceReadEndTime());
    }
    if (totalStats.getDestinationWriteStartTime() != null) {
      metadata.put("destination_write_start_time", totalStats.getDestinationWriteStartTime());
    }
    if (totalStats.getDestinationWriteEndTime() != null) {
      metadata.put("destination_write_end_time", totalStats.getDestinationWriteEndTime());
    }

    final List<FailureReason> failureReasons = failureReasonsList(attempts);
    if (!failureReasons.isEmpty()) {
      metadata.put("failure_reasons", failureReasonsListAsJson(failureReasons).toString());
      metadata.put("main_failure_reason", failureReasonAsJson(failureReasons.getFirst()).toString());
    }
    return metadata.build();
  }

  private static List<FailureReason> failureReasonsList(final List<Attempt> attempts) {
    return attempts
        .stream()
        .map(Attempt::getFailureSummary)
        .flatMap(Optional::stream)
        .map(AttemptFailureSummary::getFailures)
        .flatMap(Collection::stream)
        .sorted(Comparator.comparing(f -> {
          var timestamp = f.getTimestamp();
          return timestamp != null ? timestamp : 0L;
        }))
        .toList();
  }

  private static ArrayNode failureReasonsListAsJson(final List<FailureReason> failureReasons) {
    return Jsons.arrayNode().addAll(failureReasons
        .stream()
        .map(TrackingMetadata::failureReasonAsJson)
        .toList());
  }

  /**
   * Map a FailureReason to a string-to-object map, so it can be attached to telemetry calls.
   *
   * @param failureReason failure reason
   * @return failure reason as a string-to-object map.
   */
  public static JsonNode failureReasonAsJson(final FailureReason failureReason) {
    // we want the json to always include failureOrigin and failureType, even when they are null
    final Map<String, Object> linkedHashMap = new LinkedHashMap<>();
    linkedHashMap.put("failureOrigin", failureReason.getFailureOrigin());
    linkedHashMap.put("failureType", failureReason.getFailureType());
    linkedHashMap.put("internalMessage", failureReason.getInternalMessage());
    linkedHashMap.put("externalMessage", failureReason.getExternalMessage());
    linkedHashMap.put("metadata", failureReason.getMetadata());
    linkedHashMap.put("retryable", failureReason.getRetryable());
    linkedHashMap.put("timestamp", failureReason.getTimestamp());

    return Jsons.jsonNode(linkedHashMap);
  }

  private static String getFrequencyStringFromScheduleType(final ScheduleType scheduleType, final ScheduleData scheduleData) {
    switch (scheduleType) {
      case MANUAL -> {
        return "manual";
      }
      case BASIC_SCHEDULE -> {
        return TimeUnit.SECONDS.toMinutes(ScheduleHelpers.getIntervalInSecond(scheduleData.getBasicSchedule())) + " min";
      }
      case CRON -> {
        // TODO(https://github.com/airbytehq/airbyte/issues/2170): consider something more detailed.
        return "cron";
      }
      default -> {
        throw new RuntimeException("Unexpected schedule type");
      }
    }
  }

}

File: airbyte-persistence/job-persistence/src/main/kotlin/io/airbyte/persistence/job/tracker/TrackingMetadata.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.persistence.job.tracker

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ArrayNode
import io.airbyte.commons.json.Jsons
import io.airbyte.config.ActorDefinitionVersion
import io.airbyte.config.Attempt
import io.airbyte.config.AttemptFailureSummary
import io.airbyte.config.FailureReason
import io.airbyte.config.Job
import io.airbyte.config.ScheduleData
import io.airbyte.config.StandardDestinationDefinition
import io.airbyte.config.StandardSourceDefinition
import io.airbyte.config.StandardSync
import io.airbyte.config.helpers.ScheduleHelpers.getIntervalInSecond
import io.micronaut.core.util.StringUtils
import java.util.Optional
import java.util.concurrent.TimeUnit
import kotlin.math.roundToInt

/**
 * Helpers to fetch stats / metadata about Airbyte domain models and turn them into flat maps that
 * can be appended to telemetry calls.
 */
object TrackingMetadata {
  /**
   * Map sync metadata / stats to a string-to-object map so it can be attached to telemetry calls.
   *
   * @param standardSync connection
   * @return metadata / stats as a string-to-object map.
   */
  @JvmStatic
  fun generateSyncMetadata(standardSync: StandardSync): Map<String, Any> =
    buildMap {
      put("connection_id", standardSync.connectionId)
      put("source_id", standardSync.sourceId)
      put("destination_id", standardSync.destinationId)

      val frequencyString: String =
        if (standardSync.scheduleType != null) {
          getFrequencyStringFromScheduleType(standardSync.scheduleType, standardSync.scheduleData)
        } else if (standardSync.manual) {
          "manual"
        } else {
          val intervalInMinutes = TimeUnit.SECONDS.toMinutes(getIntervalInSecond(standardSync.schedule))
          "$intervalInMinutes min"
        }
      put("frequency", frequencyString)

      val operationCount = if (standardSync.operationIds != null) standardSync.operationIds.size else 0
      put("operation_count", operationCount)
      if (standardSync.namespaceDefinition != null) {
        put("namespace_definition", standardSync.namespaceDefinition)
      }

      val isUsingPrefix = standardSync.prefix != null && !standardSync.prefix.isBlank()
      put("table_prefix", isUsingPrefix)

      val resourceRequirements = standardSync.resourceRequirements

      if (resourceRequirements != null) {
        if (!resourceRequirements.cpuRequest.isNullOrBlank()) {
          put("sync_cpu_request", resourceRequirements.cpuRequest)
        }
        if (!resourceRequirements.cpuLimit.isNullOrBlank()) {
          put("sync_cpu_limit", resourceRequirements.cpuLimit)
        }
        if (!resourceRequirements.memoryRequest.isNullOrBlank()) {
          put("sync_memory_request", resourceRequirements.memoryRequest)
        }
        if (!resourceRequirements.memoryLimit.isNullOrBlank()) {
          put("sync_memory_limit", resourceRequirements.memoryLimit)
        }
      }
    }

  /**
   * Map destination definition metadata / stats to a string-to-object map so it can be attached to
   * telemetry calls.
   *
   * @param destinationDefinition destination definition
   * @return metadata / stats as a string-to-object map.
   */
  @JvmStatic
  fun generateDestinationDefinitionMetadata(
    destinationDefinition: StandardDestinationDefinition,
    destinationVersion: ActorDefinitionVersion,
  ): Map<String, Any> =
    buildMap {
      put("connector_destination", destinationDefinition.name)
      put("connector_destination_definition_id", destinationDefinition.destinationDefinitionId)
      putAll(generateActorDefinitionVersionMetadata("connector_destination_", destinationVersion))
    }

  /**
   * Map source definition metadata / stats to a string-to-object map so it can be attached to
   * telemetry calls.
   *
   * @param sourceDefinition source definition
   * @return metadata / stats as a string-to-object map.
   */
  @JvmStatic
  fun generateSourceDefinitionMetadata(
    sourceDefinition: StandardSourceDefinition,
    sourceVersion: ActorDefinitionVersion,
  ): Map<String, Any> =
    buildMap {
      put("connector_source", sourceDefinition.name)
      put("connector_source_definition_id", sourceDefinition.sourceDefinitionId)
      putAll(generateActorDefinitionVersionMetadata("connector_source_", sourceVersion))
    }

  private fun generateActorDefinitionVersionMetadata(
    metaPrefix: String,
    sourceVersion: ActorDefinitionVersion,
  ): Map<String, Any> =
    buildMap {
      put(metaPrefix + "docker_repository", sourceVersion.dockerRepository)
      val imageTag = sourceVersion.dockerImageTag
      if (!StringUtils.isEmpty(imageTag)) {
        put(metaPrefix + "version", imageTag)
      }
    }

  /**
   * Map job metadata / stats to a string-to-object map so it can be attached to telemetry calls.
   *
   * @param job job
   * @return metadata / stats as a string-to-object map.
   */
  @JvmStatic
  fun generateJobAttemptMetadata(job: Job?): Map<String, Any> =
    buildMap {
      // Early returns in case we're missing the relevant stats.
      if (job == null) {
        return@buildMap
      }
      val attempts = job.attempts
      if (attempts.isEmpty()) {
        return@buildMap
      }
      val lastAttempt: Attempt = attempts.last()
      if (lastAttempt.getOutput() == null || lastAttempt.getOutput().isEmpty) {
        return@buildMap
      }
      val jobOutput = lastAttempt.getOutput().get()
      if (jobOutput.sync == null) {
        return@buildMap
      }
      val syncSummary = jobOutput.sync.standardSyncSummary
      val totalStats = syncSummary.totalStats
      if (syncSummary.startTime != null) {
        put("sync_start_time", syncSummary.startTime)
      }
      if (syncSummary.endTime != null && syncSummary.startTime != null) {
        put("duration", ((syncSummary.endTime - syncSummary.startTime) / 1000.0).roundToInt())
      }
      if (syncSummary.bytesSynced != null) {
        put("volume_mb", syncSummary.bytesSynced)
      }
      if (syncSummary.recordsSynced != null) {
        put("volume_rows", syncSummary.recordsSynced)
      }
      totalStats?.let {
        if (totalStats.sourceStateMessagesEmitted != null) {
          put("count_state_messages_from_source", totalStats.sourceStateMessagesEmitted)
        }
        if (totalStats.destinationStateMessagesEmitted != null) {
          put("count_state_messages_from_destination", totalStats.destinationStateMessagesEmitted)
        }
        if (totalStats.maxSecondsBeforeSourceStateMessageEmitted != null) {
          put(
            "max_seconds_before_source_state_message_emitted",
            totalStats.maxSecondsBeforeSourceStateMessageEmitted,
          )
        }
        if (totalStats.meanSecondsBeforeSourceStateMessageEmitted != null) {
          put(
            "mean_seconds_before_source_state_message_emitted",
            totalStats.meanSecondsBeforeSourceStateMessageEmitted,
          )
        }
        if (totalStats.maxSecondsBetweenStateMessageEmittedandCommitted != null) {
          put(
            "max_seconds_between_state_message_emit_and_commit",
            totalStats.maxSecondsBetweenStateMessageEmittedandCommitted,
          )
        }
        if (totalStats.meanSecondsBetweenStateMessageEmittedandCommitted != null) {
          put(
            "mean_seconds_between_state_message_emit_and_commit",
            totalStats.meanSecondsBetweenStateMessageEmittedandCommitted,
          )
        }

        if (totalStats.replicationStartTime != null) {
          put("replication_start_time", totalStats.replicationStartTime)
        }
        if (totalStats.replicationEndTime != null) {
          put("replication_end_time", totalStats.replicationEndTime)
        }
        if (totalStats.sourceReadStartTime != null) {
          put("source_read_start_time", totalStats.sourceReadStartTime)
        }
        if (totalStats.sourceReadEndTime != null) {
          put("source_read_end_time", totalStats.sourceReadEndTime)
        }
        if (totalStats.destinationWriteStartTime != null) {
          put("destination_write_start_time", totalStats.destinationWriteStartTime)
        }
        if (totalStats.destinationWriteEndTime != null) {
          put("destination_write_end_time", totalStats.destinationWriteEndTime)
        }
      }

      val failureReasons = failureReasonsList(attempts)
      if (!failureReasons.isEmpty()) {
        put("failure_reasons", failureReasonsListAsJson(failureReasons).toString())
        put("main_failure_reason", failureReasonAsJson(failureReasons.first()).toString())
      }
    }

  private fun failureReasonsList(attempts: List<Attempt>): List<FailureReason> =
    attempts
      .stream()
      .map { obj: Attempt -> obj.getFailureSummary() }
      .flatMap { obj: Optional<AttemptFailureSummary> -> obj.stream() }
      .map { obj: AttemptFailureSummary -> obj.failures }
      .flatMap { obj: List<FailureReason> -> obj.stream() }
      .sorted(
        Comparator.comparing { f: FailureReason ->
          val timestamp = f.timestamp
          timestamp ?: 0L
        },
      ).toList()

  private fun failureReasonsListAsJson(failureReasons: List<FailureReason>): ArrayNode =
    Jsons.arrayNode().addAll(
      failureReasons
        .stream()
        .map { obj: FailureReason -> failureReasonAsJson(obj) }
        .toList(),
    )

  /**
   * Map a FailureReason to a string-to-object map, so it can be attached to telemetry calls.
   *
   * @param failureReason failure reason
   * @return failure reason as a string-to-object map.
   */
  fun failureReasonAsJson(failureReason: FailureReason): JsonNode {
    // we want the json to always include failureOrigin and failureType, even when they are null
    val linkedHashMap: MutableMap<String, Any> = LinkedHashMap()
    linkedHashMap["failureOrigin"] = failureReason.failureOrigin
    linkedHashMap["failureType"] = failureReason.failureType
    linkedHashMap["internalMessage"] = failureReason.internalMessage
    linkedHashMap["externalMessage"] = failureReason.externalMessage
    linkedHashMap["metadata"] = failureReason.metadata
    linkedHashMap["retryable"] = failureReason.retryable
    linkedHashMap["timestamp"] = failureReason.timestamp

    return Jsons.jsonNode<Map<String, Any>>(linkedHashMap)
  }

  private fun getFrequencyStringFromScheduleType(
    scheduleType: StandardSync.ScheduleType,
    scheduleData: ScheduleData?,
  ): String =
    when (scheduleType) {
      StandardSync.ScheduleType.MANUAL -> {
        "manual"
      }

      StandardSync.ScheduleType.BASIC_SCHEDULE -> {
        TimeUnit.SECONDS.toMinutes(getIntervalInSecond(scheduleData!!.basicSchedule)).toString() + " min"
      }

      StandardSync.ScheduleType.CRON -> {
        // TODO(https://github.com/airbytehq/airbyte/issues/2170): consider something more detailed.
        "cron"
      }

      else -> {
        throw RuntimeException("Unexpected schedule type")
      }
    }
}
